src/AppInstallerCLITests/Correlation.cpp

```diff
@@ -0,0 +1,159 @@
+    unsigned TrueMismatches;
+    unsigned FalseMatches;
+    unsigned FalseMismatches;
+
+    unsigned TotalCases() const
+    {
+        return TrueMatches + TrueMismatches + FalseMatches + FalseMismatches;
+    }
+};
+
+std::vector<TestCase> LoadTestData()
+{
+    // Creates test cases from the test data file.
+    // The format of the file is one case per line, each with tab separated values.
+    // Each row contains: AppId, AppName, AppPublisher, ARPDisplayName, ARPDisplayVersion, ARPPublisherName, ARPProductCode
+    // TODO: Cleanup data (e.g. bad encoding)
+    //       Add more test cases; particularly for non-matches
+    //       Consider using a different file format
+    std::ifstream testDataStream(TestCommon::TestDataFile("InputARPData.txt").GetPath());
+    REQUIRE(testDataStream);
+
+    std::vector<TestCase> testCases;
+
+    // TODO: There has to be a better way...
+    std::string line;
+    while (std::getline(testDataStream, line))
+    {
+        std::stringstream ss{ line };
+
+        TestCase testCase;
+        std::string appId;
+        std::string arpDisplayVersion;
+        std::string arpProductCode;
+        std::getline(ss, appId, '\t');
+        std::getline(ss, testCase.AppName, '\t');
+        std::getline(ss, testCase.AppPublisher, '\t');
+        std::getline(ss, testCase.ArpName, '\t');
+        std::getline(ss, arpDisplayVersion, '\t');
+        std::getline(ss, testCase.ArpPublisher, '\t');
+        std::getline(ss, arpProductCode, '\t');
+
+        testCase.IsMatch = true;
+
+        testCases.push_back(std::move(testCase));
+    }
+
+    return testCases;
+}
+
+ResultSummary EvaluateCorrelationMeasure(const ARPCorrelationMeasure& measure, const std::vector<TestCase>& cases)
+{
+    std::vector<ARPEntry> allARPEntries;
+    for (const auto& testCase : cases)
+    {
+        ARPEntry entry{ nullptr, true };
+        entry.Name = testCase.ArpName;
+        entry.Publisher = testCase.ArpPublisher;
+        entry.IsNewOrUpdated = true;
+        allARPEntries.push_back(entry);
+    }
+
+    ResultSummary result{};
+    for (const auto& testCase : cases)
+    {
+        // TODO: initialize with test data
+        Manifest manifest;
+        manifest.DefaultLocalization.Add<Localization::PackageName>(testCase.AppName);
+        manifest.DefaultLocalization.Add<Localization::Publisher>(testCase.AppPublisher);
+        manifest.Localizations.push_back(manifest.DefaultLocalization);
+
+        std::vector<ARPEntry> arpEntries;
+        ARPEntry entry{ nullptr, true };
+        entry.Name = testCase.ArpName;
+        entry.Publisher = testCase.ArpPublisher;
+        entry.IsNewOrUpdated = true;
+        arpEntries.push_back(entry);
+        // Add a couple of ARP entries as noise
+        for (size_t i = 0; i < std::min((size_t)0, allARPEntries.size()); ++i)
+        {
+            arpEntries.push_back(allARPEntries[i]);
+        }
+
+        auto match = measure.GetBestMatchForManifest(manifest, arpEntries);

```

Comment:
I think that the scoring algorithms should be only that; give the manifest and a single ARP entry and they compute a score in [0, 1.0].  Or to allow for optimization, construct them with a manifest so that they can pre-compute anything that is manifest specific, then compute a score per ARP entry.  A confidence threshold would be defined globally to standardize on the minimum for considering something a possible match.

Then the overall score would factor in other data, like new/updated entry, version matching, etc.  But that is agnostic of the name and publisher matching algorithm, and should mostly just be attempting to tiebreak or boost our confidence.

This test harness is about rating the name and publisher matching algorithm specifically, and would be a good place to also understand other metrics, like the gap between the first and second highest confidence intervals.  It having all of the scores enables that kind of investigation.

Reply:
This `GetBestMatchForManifest()` is defined in the base class. It loops over all the entries and finds the one with the highest score (that is also over the threshold). It's not defined by each algorithm; those only do what you say (assign a score).

The way I'm thinking about it, the scoring algorithms consider everything (name, version, if it is new), and the fact that we are currently only looking at name string matching is incidental. If we add scores for other things, the string matching can just be pushed one layer below these algorithm classes.

I'm making the threshold specific to each algorithm and not restricted to [0,1] because I think that is hard to normalize across multiple algorithms. Specially if we consider how to mix multiple scores into one. For example, I'm thinking that we could compare mixing the scores for name and publisher (or anything else) by a weighted sum, or by multiplying them, or by summing squares, and that would all affect the threshold and range.

Reply:
Ok, I didn't look closely at that, but I did find the base class header odd, so that is why.  I believe there may be value in being able to run multiple passes, such as:
1. Run a direct equality matcher looking for the easy cases where we actually have exact data
2. Run the existing normalized name and publisher matcher
3. Run these more complex ones we are working on now, which are slower, but better at matching

With this design, it is more complicated using a base class with implementation built in (not impossible, just a bit more convoluted).  If the matcher interface were separated out, then it would be easier to have a single orchestration class that can use one or more matchers.  Again, not impossible to do with the current base class, but I think you end up needing to wrap the functionality anyway.

The overall scoring algorithm considers everything, but I don't think that each name+publisher matching algorithm should separately consider other fields.  Not that they should be disallowed, as I could see potential value in versions occurring in the name, but generally those are separate algorithms.  If we want to consider different version matching heuristics, we should do so as its own thing.

So I guess where I'm headed is that a pure interface class makes generic matchers possible, and a single orchestration class/function allows us to combine them as we see fit.

> I'm making the threshold specific to each algorithm and not restricted to [0,1] because I think that is hard to normalize across multiple algorithms.

Unless the algorithm actually returns completely unbounded values, numbers in the range [0, 1] are always possible.  This confidence interval scoring is very common, and having a normalized range is what makes it possible to combine scores from different aspects.  Creating a standardize threshold for meaning also allows the different algorithms to be interchangeable to the code above.  If mine thinks that 0.5+ is a solid match, but yours thinks that 0.9+ is, then how can the code above combine them?  If the answer is "use the threshold", then the scoring algorithm could do the same to normalize to the standardized value.  And it can do it with knowledge of its statistics; a linear mapping may not be the best approach for all cases.

Reply:
I made some changes based on what you said (but after re-reading it, I think it doesn't exactly match what you wanted...).

I kept all of your original code for matching ARP using product code and normalized names, and then the heuristics only run if that doesn't return anything. That's intended to be part of the orchestrator you mentioned.

I added a base class for name+publisher matchers that take the strings and return a value in [0,1]. I kept the 3 sample matchers I had earlier. I think this is where you can plug in the matching algorithm you wanted to implement.

I kept the base class for the overall algorithm, and added an implementation that for now considers only the string matching. This was intended to be the orchestrator for mix-and-matching smaller matchers, but I just realized that I didn't do it the way you suggested because it doesn't allow for doing a full pass with a single matcher, then another one with other matcher if needed.

I also added implementations of that base class to the tests, which use only a single string matching algorithm. That way we can evaluate just the accuracy of those, without considering anything else we may end up adding on top in the orchestrator.

I still need to test this E2E and play with the matching algorithms.

Reply:
Yes, it is now worse for my purposes, as I need more ARP information than you are providing.  For my algorithm I want something more like:
```
struct IARPMatchConfidenceAlgorithm
{
    double Compute(const ARPEntry& entry) const;
};

struct WordsARPMatchConfidenceAlgorithm : public IARPMatchConfidenceAlgorithm
{
    WordsARPMatchConfidenceAlgorithm(const Manifest& manifest);
    double Compute(const ARPEntry& entry) const override;
};
```
Where the ARPEntry contains other information, like language, that I was going to leverage.

Reply:
I'll leave those changes for the future as we update the matching algorithm.

Reply:
Scratch that last comment; I updated it to use the interface you suggested. Ended up cleaner with it :D